src/Grpc.Net.Client/Internal/GrpcCall.cs

```diff
@@ -419,22 +446,67 @@ private void StartSend(HttpClient client, HttpRequestMessage message)
                     _deadlineTimer = new Timer(DeadlineExceeded, null, _timeout.Value, Timeout.InfiniteTimeSpan);
                 }
 
-                SendTask = SendAsync(client, message);
+                Log.StartingCall(Logger, Method.Type, request.RequestUri);
+                GrpcEventSource.Log.CallStart(Method.FullName);
+
+                var diagnosticSourceEnabled = 
+                    GrpcDiagnostics.DiagnosticListener.IsEnabled() &&
+                    GrpcDiagnostics.DiagnosticListener.IsEnabled(GrpcDiagnostics.ActivityName, request);
+
+                Activity? activity = null;
+
+                // Set activity if:
+                // 1. Diagnostic source is enabled
+                // 2. Logging is enabled
+                // 3. There is an existing activity (to enable activity propagation)
+                if (diagnosticSourceEnabled || Logger.IsEnabled(LogLevel.Critical) || Activity.Current != null)
+                {
+                    activity = new Activity(GrpcDiagnostics.ActivityName);
+                    activity.AddTag(GrpcDiagnostics.GrpcMethodTagName, Method.FullName);
+                    activity.Start();
+
+                    if (diagnosticSourceEnabled)
+                    {
+                        GrpcDiagnostics.DiagnosticListener.Write(GrpcDiagnostics.ActivityStartKey, request);
+                    }
+                }
+
+                SendTask = SendAsync(client, request);
+
+                // Wait until the call is complete
+                // TCS will be set in Dispose
+                var statusCode = await _callTcs.Task.ConfigureAwait(false);
+
+                if (statusCode != StatusCode.OK)
+                {
+                    GrpcEventSource.Log.CallFailed(statusCode);
+                }
+                GrpcEventSource.Log.CallStop();
+
+                // Activity needs to be stopped in the same execution context
+                if (activity != null)
+                {
+                    var statusText = statusCode.ToString("D");
+                    if (statusText != null)
+                    {
+                        activity.AddTag(GrpcDiagnostics.GrpcStatusCodeTagName, statusText);
+                    }
+
+                    if (diagnosticSourceEnabled)
+                    {
+                        GrpcDiagnostics.DiagnosticListener.Write(GrpcDiagnostics.ActivityStopKey, HttpResponse);
```

Comment:
if you don't want to use StopActivity(), please set end time before calling Write (check how StopActivity does it internally)

Reply:
also typically DS events are used with payload in an anonymous type , check out here https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.DiagnosticSource/src/DiagnosticSourceUsersGuide.md#payloads

this is needed to allow non-breaking addition of new payload property.

Reply:
Perf! 😄

Reply:
BTW, perf impact likely comes from the allocation, our reflection tricks keep the impact of property fetching low. 

The chance  change payloads need to change is probably very low, so I'm ok with explicit request/response

Reply:
Adding my usual rant about the DiagnosticSource usages already being very inconsistent. This would be another deviation from the official docs/recommendations. I’m all about perf, so maybe consider changing those docs if they are not practical/followed anyway?!

Reply:
nice to see you again @cwe1ss and looking forward working with you on OpenTelemetry!

there is no way to leverage payloads in a generic way - so there is always a very specific contract between instrumentation and listener. All listeners that want to stay agnostic will ignore payload.

I think docs cover it

```
CONSIDER - if you have an event that is so frequent that the performance of the logging is an important consideration, and you have only one data item and it is unlikely that you will ever have more data to pass to the event, and the data item is a normal class (not a value type) then you save some cost by simply by passing the data object directly without using an anonymous type wrapper
```

Reply:
I’ve just seen that passing in a single data object is a CONSIDER item in the docs, so they should be fine. I still don’t like the inconsistencies though 😄